vim-patch:8.1.1120

Shatur opened the pull request:
Hi all! I trying to port https://github.com/vim/vim/commit/543c9b1921d7605498b54afdef518e312f1b4515 (closes #12212). This patch just adds `readdir()` Vimscript function. I noticed that the documentation was partially updated in Neovim (mistake after merge?), but the function itself does not exists.
I updated documentation and `eval.c`. But currently I have an issue with porting ` src/evalfunc.c` to `src/nvim/eval/funcs.c`. Neovim does not have `vim_free`, `list_append_string`, `list_alloc`, `ga_init2`, `rettv_list_alloc` and `clear_tv` (used in original patch). What should I use instead? I'm not familiar with Neovim codebase, but want to help.

janlazo commented:
Refer to https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-Vim for porting Vim patches to Neovim. Check https://github.com/neovim/neovim/wiki/Merging-patches-from-upstream-Vim#code-differences to know which Vim functions to replace with Neovim equivalents. For functions like list_append_string, list_alloc, rettv_list_alloc and clear_tv, use the functions prefixed with `tv_`. For others, you have to check how other C functions in `src/nvim/eval/funcs.c` work. Neovim aborts on error so you can omit some of the error checking that Vim does.

You have to redo all commits here to something that https://github.com/neovim/neovim/blob/master/scripts/vim-patch.sh generates so that the patch can be attributed to Vim.

erw7 commented:
Regarding the porting of 8.1.1120, I think it is better to use `os_scandir`, `os_scandir_next` and `os_closedir` instead of `opendir`, `readdir` and `closedir`. I think that would eliminate the need for an implementation in `#ifdef MSWIN`.

Shatur committed:
vim-patch:8.1.1120: cannot easily get directory entry matches

Problem:    Cannot easily get directory entry matches.
Solution:   Add the readdir() function. (Yasuhiro Matsumoto, closes vim/vim#2439)
https://github.com/vim/vim/commit/543c9b1921d7605498b54afdef518e312f1b4515

Shatur95 commented:
@janlazo, thank you for the information! I missed the wiki page about porting Vim patches. I force pushed changes branch created by `vim-patch.sh`.

@erw7, thanks for the suggestion. The code looks much cleaner now.

I tested the added function locally and it works as Vim.

Shatur95 commented on src/nvim/eval/funcs.c:
```diff
@@ -6314,6 +6314,97 @@ static void f_range(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+      if (path == NULL)
+        break;
+
+      ignore = path[0] == '.' &&
+        (path[1] == NUL ||
+         (path[1] == '.' && path[2] == NUL));
```
How should I write this lines? Now it looks as in Vim, but as for this is not very readable.

Shatur95 commented on src/nvim/eval/funcs.c:
```diff
@@ -6314,6 +6314,97 @@ static void f_range(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+  ga_init(&ga, (int)sizeof(char *), 20);
+
+  if(!os_scandir(&dir, path)) {
+    smsg(_(e_notopen), path);
```
`e_notopen` corresponds to "E484: Can't open file %s" as in Vim. But as for me this is not related to directories. Should I leave it as is?

erw7 commented on src/nvim/eval/funcs.c:
```diff
@@ -6314,6 +6314,99 @@ static void f_range(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+
+      if (!ignore) {
+        ga_grow(&ga, 1);
+        ((char_u **)ga.ga_data)[ga.ga_len++] = vim_strsave((char_u *)path);
```
`vim_strsave()` is defined as follows. Therefore, if you use `xsrdup()` instead of `vim_strsave()`, you don't need to cast the `path`.

```C
char_u *vim_strsave(const char_u *string)
  FUNC_ATTR_NONNULL_RET FUNC_ATTR_MALLOC FUNC_ATTR_NONNULL_ALL
{
  return (char_u *)xstrdup((char *)string);
}
```

Ref. #459.

Shatur95 replied:
Okay! Should I also replace `char_u` with `uint8_t`?

erw7 replied:
> where only characters are used it's ok to use plain char

Therefore, I think `char` is fine.

Shatur95 replied:
Thank you, I will do.

erw7 replied:
I don't think you have to go that far.

Shatur95 replied:
Okay, I will only replace `vim_strsave`.

Shatur95 replied:
Okay :)

Shatur95 replied:
Done! `uint8_t` It was only in one place, except for `argv[0].val.v_string` and `sort_strings()`.

Shatur95 commented:
The Travis CI build was failed, but it looks like it is the CI problem.
